PR #2409 (Memory Error Cleanup) breaks with gcc/6.1.0 and c++14

I have a test included with the apollo spot check that verifies HPX.  This test uses hpx/1.2.1 which is built with gcc/6.1.0.  The test also requires c++14. 

I think there was a C++14 issues that is tracked in gcc via: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66297

It appears that the issues is addressed in 7.x but not older versions of 6.x.

Here is the error from the test:

```C++
In file included from /home/jsmiles/Development/test-tpls-build/kokkos/core/src/Kokkos_HostSpace.hpp:57:0,
                 from /home/jsmiles/Development/test-tpls-build/kokkos/core/src/Kokkos_Atomic.hpp:71,
                 from /home/jsmiles/Development/test-tpls-build/kokkos/core/src/impl/Kokkos_Spinwait.cpp:47:
/home/jsmiles/Development/test-tpls-build/kokkos/core/src/impl/Kokkos_Error.hpp:134:30: error: enclosing class of constexpr non-static member function ‘size_t Kokkos::Experimental::RawMemoryAllocationFailure::attempted_size() const’ is not a literal type
   KOKKOS_CONSTEXPR_14 size_t attempted_size() const noexcept {
                              ^~~~~~~~~~~~~~
/home/jsmiles/Development/test-tpls-build/kokkos/core/src/impl/Kokkos_Error.hpp:73:7: note: ‘Kokkos::Experimental::RawMemoryAllocationFailure’ is not literal because:
 class RawMemoryAllocationFailure : public std::bad_alloc {
       ^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/jsmiles/Development/test-tpls-build/kokkos/core/src/impl/Kokkos_Error.hpp:73:7: note:   ‘Kokkos::Experimental::RawMemoryAllocationFailure’ has a non-trivial destructor
/home/jsmiles/Development/test-tpls-build/kokkos/core/src/impl/Kokkos_Error.hpp:138:30: error: enclosing class of constexpr non-static member function ‘size_t Kokkos::Experimental::RawMemoryAllocationFailure::attempted_alignment() const’ is not a literal type
   KOKKOS_CONSTEXPR_14 size_t attempted_alignment() const noexcept {
                              ^~~~~~~~~~~~~~~~~~~
/home/jsmiles/Development/test-tpls-build/kokkos/core/src/impl/Kokkos_Error.hpp:142:43: error: enclosing class of constexpr non-static member function ‘Kokkos::Experimental::RawMemoryAllocationFailure::AllocationMechanism Kokkos::Experimental::RawMemoryAllocationFailure::allocation_mechanism() const’ is not a literal type
   KOKKOS_CONSTEXPR_14 AllocationMechanism allocation_mechanism() const
                                           ^~~~~~~~~~~~~~~~~~~~
/home/jsmiles/Development/test-tpls-build/kokkos/core/src/impl/Kokkos_Error.hpp:147:35: error: enclosing class of constexpr non-static member function ‘Kokkos::Experimental::RawMemoryAllocationFailure::FailureMode Kokkos::Experimental::RawMemoryAllocationFailure::failure_mode() const’ is not a literal type
   KOKKOS_CONSTEXPR_14 FailureMode failure_mode() const noexcept {
                                   ^~~~~~~~~~~~``

--- a/core/src/impl/Kokkos_Error.hpp
+++ b/core/src/impl/Kokkos_Error.hpp
@@ -131,22 +131,18 @@ class RawMemoryAllocationFailure : public std::bad_alloc {
   }
 
   KOKKOS_ATTRIBUTE_NODISCARD
-  KOKKOS_CONSTEXPR_14 size_t attempted_size() const noexcept {
-    return m_attempted_size;
-  }
+  size_t attempted_size() const noexcept { return m_attempted_size; }
+
   KOKKOS_ATTRIBUTE_NODISCARD
-  KOKKOS_CONSTEXPR_14 size_t attempted_alignment() const noexcept {
-    return m_attempted_alignment;
-  }
+  size_t attempted_alignment() const noexcept { return m_attempted_alignment; }
+
   KOKKOS_ATTRIBUTE_NODISCARD
-  KOKKOS_CONSTEXPR_14 AllocationMechanism allocation_mechanism() const
-      noexcept {
+  AllocationMechanism allocation_mechanism() const noexcept {
     return m_mechanism;
   }
+
   KOKKOS_ATTRIBUTE_NODISCARD
-  KOKKOS_CONSTEXPR_14 FailureMode failure_mode() const noexcept {
-    return m_failure_mode;
-  }
+  FailureMode failure_mode() const noexcept { return m_failure_mode; }
 
   void print_error_message(std::ostream &o) const;
   KOKKOS_ATTRIBUTE_NODISCARD

